catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/AsyncFileAlterationObserver.java

```diff
@@ -367,12 +402,36 @@ private void initChildEntries(AsyncFileEntry parent) throws IllegalStateExceptio
     }
   }
 
-  private void onFinish() {
+  private void onFinish(AsyncFileEntry entry) {
     synchronized (processingLock) {
-      if (processing.decrementAndGet() == 0) {
+      processing.remove(entry);
+      if (processing.isEmpty()) {
+        LOGGER.debug("All files finished processing");
         serializer.store(rootFile.getName(), rootFile);
         isProcessing = false;
       }
     }
   }
+
+  private class LogProcessing extends TimerTask {
+
+    /** Log files still in processing at scheduled intervals */
+    public void run() {
+      if (!processing.isEmpty()) {
+        String files =
+            processing.stream().map(AsyncFileEntry::getName).collect(Collectors.joining(", "));
+        LOGGER.debug("{} files being processed: {}", processing.size(), files);
+      }
+    }
+  }
+
+  @Override
+  protected void finalize() throws Throwable {
```

Comment:
❗ using finalization is not advisable. In addition, this object will never be GCed since at init time, a timer task is created and registered in the timer. Therefore, there will always be a hard reference to this object which is why it will never be GCed